processor/src/fast/mod.rs

```diff
@@ -349,294 +421,315 @@ impl FastProcessor {
-        // found, return an error.
-        match program.find_procedure_root(callee_hash) {
-            Some(callee_id) => self.execute_mast_node(callee_id, program, kernel, host).await?,
-            None => {
-                let mast_forest = host
-                    .get_mast_forest(&callee_hash)
-                    .await
-                    .ok_or_else(|| ExecutionError::dynamic_node_not_found(callee_hash, &err_ctx))?;
-
-                // We limit the parts of the program that can be called externally to procedure
-                // roots, even though MAST doesn't have that restriction.
-                let root_id = mast_forest
-                    .find_procedure_root(callee_hash)
-                    .ok_or(ExecutionError::malfored_mast_forest_in_host(callee_hash, &err_ctx))?;
-
-                self.execute_mast_node(root_id, &mast_forest, kernel, host).await?
+                let err_ctx = err_ctx!(current_forest, dyn_node, self.source_manager.clone());
+
+                // Retrieve callee hash from memory, using stack top as the memory
+                // address.
+                let callee_hash = {
+                    let mem_addr = self.stack_get(0);
+                    self.memory
+                        .read_word(self.ctx, mem_addr, self.clk, &err_ctx)
+                        .map_err(ExecutionError::MemoryError)?
+                };
+
+                // Drop the memory address from the stack. This needs to be done
+                // BEFORE saving the context,
+                // because the next instruction starts with a "shifted left" stack.
+                self.decrement_stack_size();
+
+                // For dyncall, save the context and reset it.
+                if dyn_node.is_dyncall() {
+                    self.save_context_and_truncate_stack();
+                    self.ctx = self.clk.into();
+                    self.fmp = Felt::new(FMP_MIN);
+                    self.caller_hash = callee_hash;
+                };
+
+                // push callee node
+                continuation_stack.push_node_in_phase(current_node_id, ProcessingStepPhase::Finish);
+
+                // if the callee is not in the program's MAST forest, try to find a
+                // MAST forest for it in
+                // the host (corresponding to an external library loaded in the
+                // host); if none are found, return
+                // an error.
+                match current_forest.find_procedure_root(callee_hash) {
+                    Some(callee_id) => {
+                        continuation_stack.push_node_with_decorators(callee_id);
+                    },
+                    None => {
+                        let mast_forest =
+                            host.get_mast_forest(&callee_hash).await.ok_or_else(|| {
+                                ExecutionError::dynamic_node_not_found(callee_hash, &err_ctx)
+                            })?;
```

Comment:
This doesn't seem to handle loading of the advice data from the new MAST forests (i.e., similar to how it is done in `resolve_external_node()` on line 1090 - 1092 below). We should probably encapsulate loading of MAST forests into a small helper procedure which is used both here and there.

Also, do we not need to call  `continuation_stack.push_new_forest()` here similar to how we do it on line 731 below?

Reply:
Fixed - although I didn't create the function that encapsulates yet (will give another look tomorrow). The (small) challenge is that we return different errors when loading a mast forest depending on if we're processing an external node or dyn node - so I just have to think how to make this cleanest.

Reply:
> This doesn't seem to handle loading of the advice data from the new MAST forests (i.e., similar to how it is done in resolve_external_node() on line 1090 - 1092 below).

This was also missing in `Process::execute_dyn_node()`, so I fixed it there too. I did it as a quick fix though, since we're planning on removing `Process` soon.

Reply:
> We should probably encapsulate loading of MAST forests into a small helper procedure which is used both here and there.

Now done too